cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1160 @@
+
+std::shared_ptr<arrow::Array> GetArrayFromJSON(
+    const std::shared_ptr<arrow::DataType>& type, arrow::util::string_view json) {
+    std::shared_ptr<arrow::Array> out;
+    ABORT_ON_FAILURE(arrow::ipc::internal::json::ArrayFromJSON(type, json, &out));
+  return out;
+}
+
+std::shared_ptr<arrow::RecordBatch> GetRecordBatchFromJSON(
+    const std::shared_ptr<arrow::Schema>& schema, arrow::util::string_view json) {
+  // Parse as a StructArray
+  auto struct_type = struct_(schema->fields());
+  std::shared_ptr<arrow::Array> struct_array = GetArrayFromJSON(struct_type, json);
+
+  // Convert StructArray to RecordBatch
+  return *arrow::RecordBatch::FromStructArray(struct_array);
+}
+
+std::string GetDataAsCsvString() {
+    std::string data_str = "";
+
+    data_str.append("a,b\n");
+    data_str.append("1,null\n");
+    data_str.append("2,true\n");
+    data_str.append("null,true\n");
+    data_str.append("3,false\n");
+    data_str.append("null,true\n");
+    data_str.append("4,false\n");
+    data_str.append("5,null\n");
+    data_str.append("6,false\n");
+    data_str.append("7,false\n");
+    data_str.append("8,true\n");
+
+    return data_str;
+}
+
+arrow::Status CreateDataSetFromCSVData(
+    std::shared_ptr<arrow::dataset::InMemoryDataset> &dataset) {
+    arrow::io::IOContext io_context = arrow::io::default_io_context();
+    std::shared_ptr<arrow::io::InputStream> input;
+    std::string csv_data = GetDataAsCsvString();
+    arrow::util::string_view sv = csv_data;
+    input = std::make_shared<arrow::io::BufferReader>(sv);
+
+    auto read_options = arrow::csv::ReadOptions::Defaults();
+    auto parse_options = arrow::csv::ParseOptions::Defaults();
+    auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+    // Instantiate TableReader from input stream and options
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::csv::TableReader> table_reader,
+      arrow::csv::TableReader::Make(io_context,
+                                    input,
+                                    read_options,
+                                    parse_options,
+                                    convert_options));
+
+    std::shared_ptr<arrow::csv::TableReader> reader = table_reader;
+
+    // Read table from CSV file
+    ARROW_ASSIGN_OR_RAISE(auto maybe_table,
+      reader->Read());
+    auto ds = std::make_shared<arrow::dataset::InMemoryDataset>(maybe_table);
+    dataset = std::move(ds);
+    return arrow::Status::OK();
+}
+
+std::shared_ptr<arrow::dataset::Dataset> CreateDataset() {
+    std::shared_ptr<arrow::dataset::InMemoryDataset> im_dataset;
+    ABORT_ON_FAILURE(CreateDataSetFromCSVData(im_dataset));
+    return im_dataset;
+}
+
+arrow::Status exec_plan_end_to_end_sample() {
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
```

Comment:
We shouldn't need to call this, right?

Reply:
In general, if we're calling anything labeled `internal` in an example, then something else needs fixing, I think.

Reply:
Hmm, does this function call do some registration (AddFactory) of ops associated with the dataset namespace?

Reply:
It is called already when you use a relevant API. Do not call it directly, or if you need to call it directly, then it should be fixed. We should not use internal APIs in an example that is meant for users.

Reply:
Sure, will remove this.

Reply:
/// Register dataset-based exec nodes with the exec node registry
///
/// This function must be called before using dataset ExecNode factories
ARROW_DS_EXPORT void Initialize();

this comment is in plan.h

Reply:
The challenge is that factory registration for most things happens in `arrow::compute::default_exec_factory_registry()` but the `compute` namespace should not know about the `dataset` namespace.

This is one of those things I think we can get rid of with an `engine` namespace (I created https://issues.apache.org/jira/browse/ARROW-15238 so I don't forget about this) but I don't think I'll get to that PR until after 7.0.0.

In the meantime I vote for simply not documenting but including in the examples so they compile and work.

Reply:
Ok, that seems reasonable. We can just do it once in main() then.